core/runtime/src/main/java/io/quarkiverse/operatorsdk/runtime/QuarkusFieldSelector.java

```diff
@@ -14,4 +59,97 @@ public QuarkusFieldSelector(List<Field> fields) {
+
+    static <R extends HasMetadata> Set<String> getValidFieldNamesOrNullIfDefault(Class<R> resourceClass,
+            BuildTimeConfigurationService configurationService) {
+        // first check if we have a known resource type as defined by:
+        // https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/#list-of-supported-fields
+        final var validNames = knownValidFields.get(resourceClass.getName());
+        if (validNames == null) {
+
+            if (CustomResource.class.isAssignableFrom(resourceClass)) {
+                // check if the CRD defines selectable fields
+                final var infos = configurationService.getCrdInfo().getCRDInfosFor(CRDUtils.crdNameFor(resourceClass));
+                if (infos != null) {
+                    final var info = infos.get(CRDUtils.DEFAULT_CRD_SPEC_VERSION);
+                    if (info != null) {
+                        try {
+                            final var crd = CRDUtils.loadFrom(Path.of(info.getFilePath()));
+                            return crd.getSpec().getVersions().stream()
+                                    .map(CustomResourceDefinitionVersion::getSelectableFields)
+                                    .filter(Objects::nonNull)
+                                    .flatMap(fields -> fields.stream().map(SelectableField::getJsonPath))
+                                    .collect(Collectors.toSet());
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                }
+            }
+
+            // if not a CustomResource or one that doesn't define selectable fields, return null
+            return null;
+        } else {
+            return validNames;
+        }
+    }
+
+    public static FieldSelector.Field from(String fieldSelector, Set<String> validFieldNames, Class<?> resourceClass) {
+        if (fieldSelector == null || fieldSelector.isEmpty()) {
+            return null;
+        }
+
+        final var equalsIndex = fieldSelector.indexOf('=');
+        if (equalsIndex < 0) {
+            throw new IllegalArgumentException(
+                    "Field selector must be use the <field><op><value> format where 'op' is one of '=', '==' or '!='. Was: "
+                            + fieldSelector);
+        }
+
+        var fieldNameWithPossibleBang = fieldSelector.substring(0, equalsIndex).trim();
+        if (fieldNameWithPossibleBang.isEmpty()) {
+            throw new IllegalArgumentException(
+                    "Field selector must be use the <field><op><value> format where 'op' is one of '=', '==' or '!='. Was: "
+                            + fieldSelector);
+        }
+        // check for possible ! as last character if the field is negated
+        boolean negated = false;
+        final var lastIndex = fieldNameWithPossibleBang.length() - 1;
+        if (fieldNameWithPossibleBang.indexOf('!') == lastIndex) {
```

Comment:
should it be possible to escape ! character in the field key value? Shouldn't here rather be charAt to also save the iteration through the string?